### ISSUE TITLE
Invalid credentials

### ISSUE THREAD

**Issue by MisterP64** [tellerbop/havistapool#9]
Hey guys,  I installed all codes into the custom components folder. When I install the integration into HA the system asked me as disceibed in the guideline to set account Infos as user and PW. The user (my mail address) and password are definitively correct, but the system refused to login : invalid credentials.  Anyone else who had this problem? Ha version 2021.5.0
Thanks in advance for your help
Peter

**Comment by tellerbop** [comment]
The component is currently in a non working state due to underlying service API changes. I’m working on a fix, stay tuned.

**Comment by ppenders** [comment]
Hi Tellerbop,

Looking forward to it - about to fill the pool :)

Thanks!
Perry

**Comment by tellerbop** [comment]
I've been in contact with Hayward for a long time now but I'm afraid it has been without success. They are currently not planning to implement any API for integrators to use during 2021. 
They are instead pointing us to use MODBUS RS-485 tcp/502 but at least with my installation I see no possibility to enable it.

Below is the latest email I sent. I'll get back to you when and if I get an answer but until then I'm afraid there will be no development.

_Excuse me for being blunt but I and many others are a bit frustrated about this situation.

I think Hayward/Sugar Valley (and many other companies) still misunderstand how many users look at "home automation". Home automation is not just providing an application, it's about providing integration possibilities. I have 40+ apps in my phone from companies like yours which are either required to install the appliance or to manage it. However, many actually DO provide API's for their services as well so I can live with having them installed just to install them.

But in the home automation world many users do not want another application. We want one place to control every aspect of our home, One place to receive notifications and one place to harmonize integrations.
For instance - Your app doesn't even provide notifications, something I HAD with the old API because we implemented it based on the data you provide. The user can even choose what platform to receive the notification, like Signal, Telegram, Facebook, etc.

Look at this ever growing list of integrations: Integrations - [Home Assistant (home-assistant.io) ](https://www.home-assistant.io/integrations/) or [Homey Apps](https://homey.app/en-us/apps/) This is where your spot should be.

I BOUGHT Hayward Control Station ONLY because it had an API. If it wasn't, I would have bought another brand, probably a swedish one. Now you have destroyed this functionality leaving us with a non working integration, useless application and the only option you have left us with is MODBUS. But in my case, there are NO instructions whatsoever on how to enable the interface.

You are creating angry customers by handling the situation like this. It's not customer centricity and definitely bad PR.

I and others are definitely here to help, I'm offering my time and help to provide a working integration with one of the largest home automation applications out there and I'm sure many with me will follow for other platforms - but we can't do it without your help.

Please prioritize and forward to whom it may concern._

**Comment by uspino2** [comment]
Same answer here form Sugar Valley (Hayward): there will be no API access.

In my case, I was able to install a Modbus-to-LAN interface next to the unit and I have to say it works flawlessly, and no cloud needed. But I understand it might be cumbersome for some installations. Thanks for trying. It was great while it lasted.

**Comment by mutzl** [comment]
Hayward is activly promoting the Modbus functionality in their product.
![image](https://user-images.githubusercontent.com/2855185/121490218-1b68c480-c9d5-11eb-915f-ec93c858c09c.png)

But when I asked Hayward for a documentation they just replyed with 2 lines...

_All our products include MODBUS technology RS485, so you can use it with no problem. 
You must ask your dealer about the Modbus protocol when you buy the product. 
Muchas gracias_

WTF

**Comment by tellerbop** [comment]
My dealer has NO CLUE about modbus. But I think I've managed to find out where it is. It's not obvious, it's not marked on the board but I found a clue [here ](https://tasmota.github.io/docs/NeoPool/) and will buy a RS485 to wifi adapter to test with. It's just that you pay quite a bit of money for the wifi module but you can do absolutely nothing with it :)

**Comment by mutzl** [comment]
_quite a bit of money_ ... that's what I call an _understatement_.

**Comment by mutzl** [comment]
I have a working implementation of a VistaPool integration.

But it's written in C# and therefore requires [NetDaemon](https://netdaemon.xyz/).
The code is a quick and dirty PoC and by no means production ready. But it's running here for a couple of weeks without any issues.

So if anybody is interested I can create a public repo and share the code.

**Comment by ppenders** [comment]
@mutzl That would be great! - I'd just like to get the temperature en redox readings

**Comment by olafvs** [comment]
Have you tried reverse engineering the API? I've never done such a thing, but I looked into it yesterday and I found the source and some logic for some parameters. Not everything though and I don't know my skills are sufficient.

**Comment by mutzl** [comment]
> My dealer has NO CLUE about modbus. But I think I've managed to find out where it is. It's not obvious, it's not marked on the board but I found a clue [here ](https://tasmota.github.io/docs/NeoPool/) and will buy a RS485 to wifi adapter to test with. It's just that you pay quite a bit of money for the wifi module but you can do absolutely nothing with it :)

My vendor just sent me some MODBUS documentation. 
If you want me to share that, just send me an email.

**Comment by tellerbop** [comment]
> Have you tried reverse engineering the API? I've never done such a thing, but I looked into it yesterday and I found the source and some logic for some parameters. Not everything though and I don't know my skills are sufficient.

Yes i have @mutzl has a working copy in c#. However, this is not an official API, it's querying directly against Google Firebase and out of courtesy and respect I'm asking for permission to use it.

**Comment by olafvs** [comment]
@mutzl Could you please share your integration. 

I'll send you an email for the modbus documentation, although I don't know I'll be able to use it. At least I'll have it for future reference.

**Comment by ppenders** [comment]
@olafvs could you share the mobus docs with me too?

Thanks!

Perry

**Comment by olafvs** [comment]
These are the three files I received.

[https://file.io/TkzXKKVEAIdm](https://file.io/TkzXKKVEAIdm)

[https://file.io/Xl75qO2EqN8i](https://file.io/Xl75qO2EqN8i)

[https://file.io/2fCGlSpQqcAQ](https://file.io/2fCGlSpQqcAQ)

Never used file.io before. Let me know if it doesn't work.

**Comment by rohrsh** [comment]
No luck .... file.io: "The file you requested has been deleted"

Keen to try Modbus if all other roads are dead ends. I have a USR ethernet device.

**Comment by ppenders** [comment]
Same here: "The file you requested has been deleted"

**Comment by olafvs** [comment]
Ok, Of cource I can mail them, but I want to keep the files online for future reference.

Another free service:
[https://gofile.io/d/8ZitaN](https://gofile.io/d/8ZitaN)

**Comment by ppenders** [comment]
This works - thanks!

**Comment by MisterP64** [comment]
Hey guys, obviously you're all more pros than me. Since days I try to setup Modus in HA without success. To be honest I didn't understand your conversation: I have a WIFI Module within VISTAPOOL. Does it mean, that I have to buy another component to get things running? I downloaded the MODBUS files but I can't use them because a lack of knowledge. If possible it would be great to have a step by step introduction how to use all the stuff. Thanks a lot Peter

**Comment by ppenders** [comment]
@MisterP64 - I got it to work after fiddling around for a hours and are able to read temp etc. I'd be happy to help out if needed however, you do need to purchase a modbus interface to connect to either "display" or "external" i used this one: https://www.amazon.nl/gp/product/B08Q4MNT4P/ref=ppx_yo_dt_b_asin_title_o01_s00?ie=UTF8&psc=1

**Comment by tellerbop** [comment]
> Hey guys, obviously you're all more pros than me. Since days I try to setup Modus in HA without success. To be honest I didn't understand your conversation: I have a WIFI Module within VISTAPOOL. Does it mean, that I have to buy another component to get things running? I downloaded the MODBUS files but I can't use them because a lack of knowledge. If possible it would be great to have a step by step introduction how to use all the stuff. Thanks a lot Peter

Yes you need to buy a separate component as the wifi module included is not modbus enabled. I’ve commented this fact to Hayward but they just keep ignoring the stupidity in this. 😂 
I’ve ordered some stuff from China to build my own module with a local rest api to avoid the modbus implementation - just might sell it when done as Hayward clearly does not know what modern home automation is. 😉

**Comment by MisterP64** [comment]
@ppenders and tellerbop thanks for the quick response. I'll try the modus thing, but I fear that I will not succeed in HA modes setup. I cannot understand the companies like sugar valley. same happened with Mercedes me and others. They all want to do their thing and they are not customer oriented. Does someone have a config example of the modes in HA, who would like to share it? Regards Peter

**Comment by ppenders** [comment]
@MisterP64 this topic can get you going and has a lot of examples to get started: https://community.home-assistant.io/t/vistapool-integration/128695/10

As mentioned before - you will need additional hardware.

**Comment by fdebrus** [comment]
The new release of Hayward uses firestore from google, I'm not familiar with this but I was able to reverse engineer the website to get pool value. So far, I've not been able to send command. if anyone is familiar with firestore, we could surely get it done. 

sharing my code to read pool data, it's written for openhab, but you can understand what it's doing. You can get the API key from the developer console on website poolwatch.hayward.fr

> 
	// Verify the user account and get id and refresk token
	var jsonResponse = sendHttpPostRequest('https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyPassword?key='+Hayward_APIKey, 'application/json', '{email: "'+Hayward_User+'", password: "'+Hayward_Passwd+'", returnSecureToken: true}',5000)
	val idToken = transform("JSONPATH","$.idToken", jsonResponse)
	val refreshToken = transform("JSONPATH","$.refreshToken", jsonResponse)
	Hayward_refreshToken.postUpdate(refreshToken)

	// Using the received if token get the access token and user id
	jsonResponse = sendHttpPostRequest('https://www.googleapis.com/identitytoolkit/v3/relyingparty/getAccountInfo?key='+Hayward_APIKey, 'application/json', '{idToken:"'+idToken+'"}',5000)
	jsonResponse = sendHttpPostRequest('https://securetoken.googleapis.com/v1/token?key='+Hayward_APIKey, 'application/json', '{grant_type:"refresh_token", refresh_token:"'+refreshToken+'"}', 5000)
	val access_token = transform("JSONPATH","$.access_token", jsonResponse )
	val user_id = transform("JSONPATH","$.user_id", jsonResponse)
	Hayward_accessToken.postUpdate(access_token)

	// Using the access token get the pool id
	jsonResponse = queryAPI.apply('https://firestore.googleapis.com/v1beta1/projects/hayward-europe/databases/(default)/documents/users/'+user_id, readResponse)
	postUpdate('Hayward_poolID', transform("JSONPATH","$.fields.pools.arrayValue.values[0].stringValue", jsonResponse))

	// READ THE POOL DATA
	jsonResponse = queryAPI.apply('https://firestore.googleapis.com/v1beta1/projects/hayward-europe/databases/(default)/documents/pools/'+Hayward_poolID.state.toString, readResponse)

**Comment by MisterP64** [comment]
I bought the additional hardware two days after you mentioned that I reedit. Since weeks now I try to get this modes thing started. I just have problems to reach the additional hardware ppenders recommended to me. The thing is that I have no clue how to set up the initial things in HA. The component is shown in my WIFI setting as EW 11_..., so I am sure to have a connection with it. Attached you can see the Logs of my system.
thanks for your support
![Bildschirmfoto 2021-07-15 um 22 55 02](https://user-images.githubusercontent.com/71771608/125856932-2e8f78d2-e9c2-4120-9532-030eda46257c.png)

**Comment by MisterP64** [comment]
Hey guys, I changed the IP of the Elfin EW 11 into a new IP. Now I get a TIME OUT ; the original IP was not the classical one (10.10.100.---) I have access to the Elfin via WIFI and Router, can modify the settings, but HA sends this logfile:
![Bildschirmfoto 2021-07-17 um 10 03 41](https://user-images.githubusercontent.com/71771608/126030583-714f2679-b44d-4836-8ff7-a6f4b082c660.png)
Do you have any idea what went wrong

**Comment by MisterP64** [comment]
Hello, meanwhile I suceeded in setting up modbus and establishing a connection to the Elfin EW 11 (I checked it with the hiflying app tcpudpdgb on their Website). Nevertheless HA doesn't detect any sensor in Vistapool. Attached my config. Could you see any mistake (The X in the IP are set only in this post, real IP is set in config).  The lovelace card gave the response: sensor.mbf_measure_temperature not available. Thanks for your support
![Bildschirmfoto 2021-07-30 um 09 50 55](https://user-images.githubusercontent.com/71771608/127620338-367d0c4f-5381-4042-ae14-658af5d34489.png)
![Bildschirmfoto 2021-07-30 um 09 52 04](https://user-images.githubusercontent.com/71771608/127620341-61e41a11-a675-47f7-b47c-33f1445fc328.png)

**Comment by MisterP64** [comment]
first picture is Lovelace, second one is config folder within HA (I didn't make it clear in my last post, sorry).
